File: lua/problem_70.lua
local M = {}

local function digit_sum(n)
    local s = 0
    while n > 0 do s = s + (n % 10); n = n // 10 end
    return s
end

function M.nth_perfect_number(n)
    local count, num = 0, 0
    while count < n do
        num = num + 1
        if digit_sum(num) == 10 then count = count + 1 end
    end
    return num
end

function M.problem70()
    print("Problem 70:")
    assert(M.nth_perfect_number(1) == 19)
    assert(M.nth_perfect_number(2) == 28)
    assert(M.nth_perfect_number(3) == 37)
    print("All Tests Passed")
end

return M

File: c/problem_70.c
#include "problem_70.h"
#include <stdio.h>
#include <assert.h>

static int digit_sum(int n){ int s=0; while(n>0){s+=n%10;n/=10;} return s; }

int nth_perfect_number(int n) {
    int count=0, num=0;
    while(count<n){ num++; if(digit_sum(num)==10) count++; }
    return num;
}

void problem70(void) {
    printf("Problem 70:\n");
    assert(nth_perfect_number(1)==19);
    assert(nth_perfect_number(2)==28);
    assert(nth_perfect_number(3)==37);
    printf("All Tests Passed\n");
}
